sdks/java/core/src/main/java/org/apache/beam/sdk/testing/SourceTestUtils.java

```diff
@@ -673,4 +680,93 @@ public SplitAtFractionResult(int numPrimaryItems, int numResidualItems) {
         numItemsToReadBeforeSplitting, fraction, options);
     return (res.numResidualItems > 0);
   }
+
+  /**
+   * Returns an equivalent unsplittable {@code BoundedSource<T>}.
+   */
+  public static <T> BoundedSource<T> toUnsplittableSource(BoundedSource<T> boundedSource) {
+    return new UnsplittableSource<>(boundedSource);
+  }
+
+  private static class UnsplittableSource<T> extends BoundedSource<T> {
+
+    private final BoundedSource<T> boundedSource;
+
+    private  UnsplittableSource(BoundedSource<T> boundedSource) {
+      this.boundedSource = checkNotNull(boundedSource, "boundedSource");
+    }
+
+    @Override
+    public List<? extends BoundedSource<T>> splitIntoBundles(
+        long desiredBundleSizeBytes, PipelineOptions options) throws Exception {
+      return ImmutableList.of(this);
+    }
+
+    @Override
+    public long getEstimatedSizeBytes(PipelineOptions options) throws Exception {
+      return boundedSource.getEstimatedSizeBytes(options);
+    }
+
+    @Override
+    public boolean producesSortedKeys(PipelineOptions options) throws Exception {
+      return boundedSource.producesSortedKeys(options);
+    }
+
+    @Override
+    public BoundedReader<T> createReader(PipelineOptions options) throws IOException {
+      return new UnsplittableReader<>(boundedSource, boundedSource.createReader(options));
+    }
+
+    @Override
+    public void validate() {
+      boundedSource.validate();
+    }
+
+    @Override
+    public Coder<T> getDefaultOutputCoder() {
+      return boundedSource.getDefaultOutputCoder();
+    }
+
+    private static class UnsplittableReader<T> extends BoundedReader<T> {
+
+      private final BoundedSource<T> boundedSource;
+      private final BoundedReader<T> boundedReader;
+
+      private UnsplittableReader(BoundedSource<T> boundedSource, BoundedReader<T> boundedReader) {
+        this.boundedSource = checkNotNull(boundedSource, "boundedSource");
+        this.boundedReader = checkNotNull(boundedReader, "boundedReader");
+      }
+
+      @Override
+      public BoundedSource<T> getCurrentSource() {
+        return boundedSource;
+      }
+
+      @Override
+      public boolean start() throws IOException {
+        return boundedReader.start();
+      }
+
+      @Override
+      public boolean advance() throws IOException {
+        return boundedReader.advance();
+      }
+
+      @Override
+      public T getCurrent() throws NoSuchElementException {
+        return boundedReader.getCurrent();
```

Comment:
add unit tests that all methods are exposed correctly and that it rejects both initial splitting and dynamic splitting.